### PR TITLE
feat(api): load adapter and load/move labware onto adapters in PAPI

### PR DESCRIPTION
**Pull request by jbleon95**
# Overview

Closes RLAB-343.

This PR adds PAPI support for the labware adapter split work. This is implemented in three main ways in the public API:
- The ability to load adapters (which are any labware that has `"adapter"` in `allowedRoles` in its Labware Definition) via `ProtocolContext` and `ModuleContext` `load_adapter` and `load_adapter_from_definition`.
- The ability to load labware on top of compatible labware objects (this includes adapters) via a `Labware.load_labware` and `Labware.load_labware_from_definition`. Labware that exists on top of a `Labware` object can be accessed via `Labware.child` as a corollary to `Labware.parent`. Loading an adapter underneath a compatabile labware can also be done via the added `adapter=` keyword arg for protocol context and module `load_labware`.
- The ability to move labware on top of another labware via `move_labware`, both manually and with the gripper. This is done by giving the `Labware` object as the `new_location`

A usable protocol demonstrating the new api:
```
metadata = {
    'protocolName': 'PAPI Labware Split Demonstration',
}

requirements = {
    "robotType": "OT-3",
    "apiLevel": "2.15"
}


def run(context):
    # Loading pipette and tiprack and module
    tiprack = context.load_labware('opentrons_flex_96_tiprack_50ul', location="D3")
    pipette = context.load_instrument("p50_single_gen3", mount="right", tip_racks=[tiprack])
    heater_shaker = context.load_module("heaterShakerModuleV1", location="D1")
    heater_shaker.open_labware_latch()

    # Adapters can be loaded directly onto the deck or a module
    flat_adapter = context.load_adapter("opentrons_universal_flat_adapter", location="C3")
    pcr_adapter = heater_shaker.load_adapter("opentrons_96_pcr_adapter")

    # Labware can then be loaded onto the adapter 
    nest_plate = pcr_adapter.load_labware("nest_96_wellplate_100ul_pcr_full_skirt")

    # Alternatively, the adapter and labware can be loaded in one line, this is also possible on a module.
    # Note that via this method, if you specify a namespace it will use that for both the labware and adapter,
    # but this is not true with version, to allow the app to pick the most recent adapter definition if necessary
    armadillo_plate = context.load_labware(
        load_name="armadillo_96_wellplate_200ul_pcr_full_skirt",
        location="C1",
        adapter="opentrons_96_pcr_adapter"
    )
    # The adapter labware object can then be accessed via .parent
    other_pcr_adapter = armadillo_plate.parent
    # And via the adapter labware object, the labware on top can also be accessed via .xalabware
    assert other_pcr_adapter.child is armadillo_plate


    # Labware can be moved from an adapter
    context.move_labware(armadillo_plate, new_location="B1", use_gripper=True)

    # Labware property will be updated to reflect nothing sitting on it, along with parent
    assert other_pcr_adapter.child is None
    assert armadillo_plate.parent == "B1"

    # Labware can also be moved to an adapter
    context.move_labware(nest_plate, new_location=other_pcr_adapter, use_gripper=True)
    assert nest_plate.parent is other_pcr_adapter
    assert other_pcr_adapter.child is nest_plate
```

# Test Plan

The above protocol was tested to be able to analyzed, along with other protocols tested on the robot to ensure that the new API works as expected. Since the gripper movement logic is all done via the engine, testing done for #12928 is applicable to the logic here.

# Changelog

- Added `load_adapter` and ``load_adapter_from_definition` methods to `ProtocolContext` and `ModuleContext`s
- Added a `load_labware` and `load_labware_from_definition` to `Labware`, allowing compatible labware to be loaded on top of that labware object
- Added an `adapter` keyword to `load_labware` for `ProtocolContext` and `ModuleContext` to allow one line loading of split adapter and labware
- `ProtocolContext.move_labware` can now accept `Labware` objects as the `new_location` destination. Moving from on top of a labware is also supported.
- Added a `Labware.child` property to return the `Labware` sitting on that labware, or `None` if there isn't any
- Added an `is_adapter` property to `Labware`
- Updated `Labware.parent` to return a `Labware` object if that is the labware's location
- Updated `validate_takes_liquid` to block liquid handling commands from working directly on adapters
- Updated ProtocolEngine `validate_liquid_allowed_in_labware` to also block liquid handling commands for adapters
- Fixed oversight where `ProtocolContext.load_labware_from_definition` didn't accept `OFF_DECK` as a location
- Fixed oversight where `ModuleContext.load_labware` default `version` keyword arg was `1` instead of `None`, which allows us to programmatically select the latest definition version.

# Review requests

The usual code and API quality, but I especially want attention paid to the doc strings I've added and updated. The big question I have is how much to directly refer to labware as "adapters` throughout the code base. 

In some ways they are sort of like tip racks, in that they are `Labware` objects but have additional restrictions. For instance, adapters themselves cannot be moved with the gripper and they don't work with liquid handling commands. But unlike tip racks, they are currently implemented with their own bespoke load method that will fail if given a non-adapter definition (and `load_labware` likewise will fail if given an adapter definition)

# Risk assessment

Medium, this isn't changing any major logic for existing load labware, but it is introducing to PAPI a new location for labware and there may be some places I missed where that assumption that labware can only be on module or deck slot, though I've been as thorough as possible looking through the code base. There will need to be some frontend work for the app and PD to fully support this feature, but this PR does not break any normal use with the app, though loading labware onto labware/adapters will cause LPC to gracefully crash (programmatically).

### CODE REVIEW

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 1]
```diff
@@ -340,14 +341,28 @@ def load_labware(
 
         :param version: The version of the labware definition. You should normally
             leave this unspecified to let the implementation choose a good default.
+        :param adapter: An load name of an adapter to load the labware on top of. The adapter
+            will be loaded from the same given namespace, but version will be automatically chosen.
```
Comment (sanni-t) [inline api/src/opentrons/protocol_api/protocol_context.py]:
```suggestion
        :param adapter: Load name of an adapter to load the labware on top of. The adapter
            will be loaded from the same given namespace, but version will be automatically chosen.
```

File: api/src/opentrons/protocol_engine/errors/exceptions.py [hunk 2]
```diff
@@ -158,6 +158,19 @@ def __init__(
...
+    def __init__(
+        self,
+        message: Optional[str] = None,
+        details: Optional[Dict[str, Any]] = None,
+        wrapping: Optional[Sequence[EnumeratedError]] = None,
+    ) -> None:
+        """Build a LabwareNotLoadedOnLabwareError."""
+        super().__init__(ErrorCodes.GENERAL_ERROR, message, details, wrapping)
+
+
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/errors/exceptions.py]: Is there a reason why we can't use `LabwareNotLoadedError` for such an error case? It doesn't seem like we raise this error when the labware we are trying to access exists but it's not on top of the labware it should be. I guess the error name is a bit confusing. No strong opinions here, just trying to understand why this error (as well as `LabwareNotLoadedOnModuleError`) was necessary.

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 3]
```diff
@@ -504,16 +568,33 @@ def get_labware_location(
...
             return ModuleLocation(moduleId=location.module_id)
         elif location is OffDeckType.OFF_DECK:
             return OFF_DECK_LOCATION
         elif isinstance(location, DeckSlotName):
             return DeckSlotLocation(slotName=location)
+        elif isinstance(location, LabwareCore):
+            return OnLabwareLocation(labwareId=location.labware_id)
+
+    @staticmethod
+    def _convert_non_stacked_location(
```
Comment (jbleon95) [inline api/src/opentrons/protocol_api/core/engine/protocol.py]: Look into refactoring this to remove duplicated code

File: api/src/opentrons/protocol_api/labware.py [hunk 4]
```diff
@@ -442,6 +445,71 @@ def magdeck_engage_height(self) -> Optional[float]:
         else:
             return p["magneticModuleEngageHeight"]
 
+    @property  # type: ignore[misc]
+    @requires_version(2, 15)
+    def labware(self) -> Optional[Labware]:
```
Comment (jbleon95) [inline api/src/opentrons/protocol_api/labware.py]: maybe rename?

File: api/src/opentrons/protocols/api_support/instrument.py [hunk 5]
```diff
@@ -100,7 +100,9 @@ def validate_tiprack(
             )
 
 
-def validate_takes_liquid(location: types.Location, reject_module: bool) -> None:
+def validate_takes_liquid(
+    location: types.Location, reject_module: bool, check_for_adapter: bool
```
Comment (jbleon95) [inline api/src/opentrons/protocols/api_support/instrument.py]: update docstrings with arguments

File: api/src/opentrons/protocol_api/instrument_context.py [hunk 6]
```diff
@@ -315,6 +316,7 @@ def dispense(
             instrument.validate_takes_liquid(
                 location=move_to_location,
                 reject_module=self.api_version >= APIVersion(2, 13),
+                check_for_adapter=self.api_version >= APIVersion(2, 15),
```
Comment (jbleon95) [inline api/src/opentrons/protocol_api/instrument_context.py]: leave a comment about potential ability to do a 2.14 load labware of adapter definition and then be able to aspirate/dispense to adapter

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 7]
```diff
@@ -472,8 +578,10 @@ def move_labware(
                 f"Expected labware of type 'Labware' but got {type(labware)}."
             )
 
-        location: Union[ModuleCore, OffDeckType, DeckSlotName]
-        if isinstance(new_location, ModuleContext):
+        location: Union[ModuleCore, LabwareCore, OffDeckType, DeckSlotName]
+        if isinstance(new_location, Labware):
+            location = new_location._core
+        elif isinstance(new_location, ModuleContext):
```
Comment (jbleon95) [inline api/src/opentrons/protocol_api/protocol_context.py]: these two lines can be combined

File: api/tests/opentrons/conftest.py [hunk 8]
```diff
@@ -283,7 +283,7 @@ def _make_ot3_pe_ctx(
     """Return a ProtocolContext configured for an OT-3 and backed by Protocol Engine."""
     with protocol_engine_in_thread(hardware=hardware) as (engine, loop):
         yield create_protocol_context(
-            api_version=APIVersion(2, 14),
+            api_version=APIVersion(2, 15),
```
Comment (jbleon95) [inline api/tests/opentrons/conftest.py]: make this MAX_SUPPORTED_API

File: api/src/opentrons/protocol_api/labware.py (hunk 4 above)
Comment (ecormany) [reply api/src/opentrons/protocol_api/labware.py]: `child` or `child_labware`?

File: api/src/opentrons/protocol_api/labware.py [hunk 9]
```diff
@@ -442,6 +445,71 @@ def magdeck_engage_height(self) -> Optional[float]:
...
+    def load_labware_from_definition(
+        self, definition: LabwareDefinition, label: Optional[str] = None
+    ) -> Labware:
+        """Load a labware onto the module using an inline definition.
+
+        :param definition: The labware definition.
+        :param str label: An optional special name to give the labware. If
+                          specified, this is the name the labware will appear
+                          as in the run log and the calibration view in the
+                          Opentrons app.
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/labware.py]:
```suggestion
                          Opentrons App.
```
🧐

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 10]
```diff
@@ -198,6 +213,58 @@ def load_labware_by_name(
...
+        version: Optional[int] = None,
+    ) -> Labware:
+        """Load an adapter onto the module using its load parameters.
+
+        The parameters of this function behave like those of
+        :py:obj:`ProtocolContext.load_adapter` (which loads adapters directly
+        onto the deck). Note that the parameter ``name`` here corresponds to
+        ``load_name`` on the ``ProtocolContext`` function.
+
+        :returns: The initialized and loaded labware object.
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/module_contexts.py]: `adapter object`?

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 11]
```diff
@@ -386,6 +401,95 @@ def load_labware_by_name(
...
+        self,
+        adapter_def: "LabwareDefinition",
+        location: Union[DeckLocation, OffDeckType],
+    ) -> Labware:
+        """Specify the presence of an adapter on the deck.
+
+        This function loads the adapter definition specified by `adapter_def`
+        to the location specified by `location`.
+
+        :param adapter_def: The adapter's labware definition to load
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/protocol_context.py]:
```suggestion
        :param adapter_def: The adapter's labware definition.
```
maybe

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 12]
```diff
@@ -386,6 +401,95 @@ def load_labware_by_name(
...
+        self,
+        load_name: str,
+        location: Union[DeckLocation, OffDeckType],
+        namespace: Optional[str] = None,
+        version: Optional[int] = None,
+    ) -> Labware:
+        """Load an adapter onto a location.
+
+        For adapters already defined by Opentrons, this is a convenient way
+        to collapse the two stages of adapters initialization (creating
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/protocol_context.py]:
```suggestion
        to collapse the two stages of adapter initialization (creating
```

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 13]
```diff
@@ -386,6 +401,95 @@ def load_labware_by_name(
...
+    @requires_version(2, 15)
+    def load_adapter_from_definition(
+        self,
+        adapter_def: "LabwareDefinition",
+        location: Union[DeckLocation, OffDeckType],
+    ) -> Labware:
+        """Specify the presence of an adapter on the deck.
+
+        This function loads the adapter definition specified by `adapter_def`
+        to the location specified by `location`.
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/protocol_context.py]:
```suggestion
        This function loads the adapter definition specified by ``adapter_def``
        to the location specified by ``location``.
```
arr ess teeeeeeee

**Review by ecormany — DISMISSED** [review]
A few typos and tweaks for docstrings that publish to the PAPI docs site.

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 14]
```diff
@@ -193,12 +198,59 @@ def load_labware(
...
 
+    def load_adapter(
+        self,
+        load_name: str,
+        location: Union[DeckSlotName, ModuleCore, NonConnectedModuleCore, OffDeckType],
+        namespace: Optional[str],
+        version: Optional[int],
+    ) -> LabwareCore:
+        """Load an adapter using its identifying parameters"""
+        load_location = self._convert_non_stacked_location(location=location)
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/engine/protocol.py]: sorry for the nip but can we call this `_get_non_stacked_location`? this name is very confusing to me

File: api/src/opentrons/protocol_api/labware.py [hunk 15]
```diff
@@ -442,6 +445,71 @@ def magdeck_engage_height(self) -> Optional[float]:
         else:
             return p["magneticModuleEngageHeight"]
 
+    @property  # type: ignore[misc]
+    @requires_version(2, 15)
+    def child_labware(self) -> Optional[Labware]:
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/labware.py]: I feel like `.labware` makes more sense in being consistent with ModuleContext. what if we call in `.child`? I feel like this makes more sense in the relationship with `.parent` and is also being used in other common languages

File: api/src/opentrons/protocol_engine/state/labware.py [hunk 16]
```diff
@@ -259,6 +260,18 @@ def get_id_by_module(self, module_id: str) -> str:
             "There is no labware loaded on this Module"
         )
 
+    def get_id_by_labware(self, labware_id: str) -> str:
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/labware.py:263]: should this be `get_loaded_labware_id` ? not critical we can leave as is or `get_loaded_labware_id_by_id`

File: api/src/opentrons/protocols/api_support/instrument.py [hunk 17]
```diff
@@ -100,11 +100,15 @@ def validate_tiprack(
...
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocols/api_support/instrument.py]: sorry one more nitpick! can we change `check_for_adapter` to `reject_adapter` for consistent?

File: api/tests/opentrons/protocol_api/core/engine/test_labware_core.py [hunk 18]
```diff
@@ -173,6 +174,21 @@ def test_is_tip_rack(subject: LabwareCore) -> None:
     assert result is True
 
 
+@pytest.mark.parametrize(
```
Comment (TamarZanzouri) [inline api/tests/opentrons/protocol_api/core/engine/test_labware_core.py:177]: lets add one more that is not an adapter?

**Review by TamarZanzouri — CHANGES_REQUESTED** [review]
beautiful work! I am ready to approve this! a few small suggestions but this is good to go!

**Review by TamarZanzouri — APPROVED** [review]
Thank you for addressing the changes. love it!

**Pull request merged**